[refactor] Move some net_processing globals into PeerManagerImpl

ajtowns opened the pull request:
Turns some globals into member variables, and simplifies the parameter list for some of net_processing's internal functions. Mostly just serves as a code cleanup at this point.

ajtowns committed:
net_processing: move some globals into PeerManagerImpl

nSyncStarted, mapBlockSource, g_wtxid_relay_peers,
g_outbound_peers_with_protect_from_disconnect were all only used by
PeerManagerImpl methods already.

ajtowns committed:
net_processing: move AlreadyHaveTx into PeerManageImpl

Allows making recentRejects and g_recent_confirmed_transactions members
rather than globals.

ajtowns committed:
net_processing: move MarkBlockAs*, TipMayBeStale, FindNextBlocksToDL to PeerManagerImpl

Allows converting mapBlocksInFlight and g_last_tip_update from globals
to member variables.

ajtowns committed:
net_processing: move FindTxForGetData and ProcessGetData to PeerManagerImpl

Allows making mapRelay and vRelayExpiration members rather than globals.

ajtowns committed:
net_processing: simplify AlreadyHaveTx args

No need to pass mempool to PeerManagerImpl methods.

ajtowns committed:
net_processing: simplify PeerManageImpl method args

No need to pass mempool to MarkBlockAsInFlight, or consensusParams to
TipMayBeStale or FindNextBlocksToDownload.

ajtowns committed:
net_processing: simplify ProcessGetData and FindTxForGetData args

No need to pass mempool or connman to PeerManagerImpl methods.

ajtowns committed:
net_processing: move MaybeSetPeerAsAnnouncingHeadersAndIDs into PeerManagerImpl

Allows making lNodesAnnouncingHeaderAndIDs and
nPeersWithValidatedDownloads member vars instead of globals.

ajtowns committed:
net_processing: simplify MaybeSetPeerAsAnnouncingHeaderAndIDs args

No need to pass connman to PeerManagerImpl methods.

ajtowns commented:
Next set of patches from #20758 -- should be straightforward to review hopefully.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -345,10 +353,7 @@ class PeerManagerImpl final : public PeerManager
 
-namespace {
     /** Number of nodes with fSyncStarted. */
     int nSyncStarted GUARDED_BY(cs_main) = 0;
```
You're renaming the other data members as you move them into `PeerManagerImpl`. Perhaps rename this to `m_sync_started`?

ajtowns replied:
I'm only renaming the ones that are currently marked `g_` since that would be wrong once they're no longer global. I'm avoiding renaming the others because that's unnecessary churn.

I think all of them are `GUARDED_BY(cs_main)` so renaming them at the same time as switching them to a new mutex would make sense -- that's the point at which you need to be reviewing every place the variable is touched to ensure the locking makes sense, so having every place the variable is touched show up in the diff because it's getting renamed would make sense.

jnewbery replied:
ok!

jnewbery commented on src/net_processing.cpp:
```diff
@@ -360,6 +365,14 @@ namespace {
      */
     std::map<uint256, std::pair<NodeId, bool>> mapBlockSource GUARDED_BY(cs_main);
```
Perhaps rename this to `m_block_sources`?

jnewbery commented on src/net_processing.cpp:
```diff
@@ -402,42 +415,44 @@ namespace {
-
-    /** Number of peers from which we're downloading blocks. */
-    int nPeersWithValidatedDownloads GUARDED_BY(cs_main) = 0;
+    Mutex m_cs_recent_confirmed_transactions;
```
Since you're already renaming this, consider dropping the outdated cs naming. Perhaps `m_recent_confirmed_transactions_mutex`?

ajtowns replied:
Sure, makes sense.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -402,42 +415,44 @@ namespace {
 
-    /** Number of peers with wtxid relay. */
-    int g_wtxid_relay_peers GUARDED_BY(cs_main) = 0;
+    bool MarkBlockAsReceived(const uint256& hash) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+    bool MarkBlockAsInFlight(NodeId nodeid, const uint256& hash, const CBlockIndex* pindex = nullptr, std::list<QueuedBlock>::iterator** pit = nullptr) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+    bool TipMayBeStale() EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+    void FindNextBlocksToDownload(NodeId nodeid, unsigned int count, std::vector<const CBlockIndex*>& vBlocks, NodeId& nodeStaller) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
Move the function comments to the declarations now that these are declared separately from the function definition.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -750,13 +765,14 @@ static void MaybeSetPeerAsAnnouncingHeaderAndIDs(NodeId nodeid, CConnman& connma
-    if (g_last_tip_update == 0) {
-        g_last_tip_update = GetTime();
+    const Consensus::Params& consensusParams = m_chainparams.GetConsensus();
+    if (m_last_tip_update == 0) {
+        m_last_tip_update = GetTime();
```
Now that `m_last_tip_update` isn't a global and can be initialized at the time that `PeerManagerImpl` is constructed, consider initializing it to `GetTime()` in the initializer list and removing this special casing.

ajtowns replied:
That seems like a good idea -- I think it'd let you mark `TipMayBeStale` as a const method as well; but I think it goes beyond the mechanical code shuffling that the rest of this PR does.

jnewbery replied:
Yep, fine to leave this. There's plenty more to be cleaned up after this PR.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -402,42 +415,44 @@ namespace {
-    std::atomic<int64_t> g_last_tip_update(0);
+    std::atomic<int64_t> m_last_tip_update{0};
+
+    CTransactionRef FindTxForGetData(const CNode& peer, const GenTxid& gtxid, const std::chrono::seconds mempool_req, const std::chrono::seconds now) LOCKS_EXCLUDED(cs_main);
```
Move function comment to the declaration.

jnewbery reviewed (commented):
Code review ACK a3ef94957be28d1b94ee0a34e015b39e60895e0d

This seems fine, but since this is mostly a code cleanup PR, it'd be nice to tidy this up a bit further while you're touching all these lines:

- you rename some of the variables you're moving to the standard `m_` convention, but leave others unchanged. Perhaps move all of them without changing the naming, and then finish this branch with a single scripted diff which changes everything to use standard naming?
- The `PeerManagerImpl` declaration mixes up the data members and member functions. I think it's much easier to read if the data members are grouped first and then the member functions. Perhaps add a final move-only commit that groups data members and member functions?
- Some of the member function comments have been left attached to the definitions rather than moved to the declaration.

ajtowns commented:
Rebased, addressed some of @jnewbery's comments. There's still more stuff coming from #20758 so will leave reorganising PeerManagerImpl declaration until later.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -402,42 +415,65 @@ namespace {
 
-    /** Stack of nodes which we have set to announce using compact blocks */
-    std::list<NodeId> lNodesAnnouncingHeaderAndIDs GUARDED_BY(cs_main);
+    /* Returns a bool indicating whether we requested this block.
```
This (and the comment for `MarkBlockAsInFlight()`) won't get picked up by doxygen without the `/**` comment marker (https://www.doxygen.nl/manual/docblocks.html)

ajtowns replied:
Added a todo commit in #20758 to track this

jnewbery reviewed (approved):
Code review ACK 6452190841f8da1cdaf899d064974136ab37659f

One minor suggestion inline

ariard commented on src/net_processing.cpp:
```diff
@@ -362,10 +362,9 @@ class PeerManagerImpl final : public PeerManager
-} // namespace
 
-namespace {
+    bool AlreadyHaveTx(const GenTxid& gtxid, const CTxMemPool& mempool) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
eeac506

I don't know if we have code guidelines on this but maybe move this method at the bottom of the other ones instead of among variables members. 

Also maybe add a small comment "Check if we have already seen this transaction by looking up its identifier among reject filter, orphan pool, recent confirmations and mempool" ?

jnewbery replied:
See discussion here: https://github.com/bitcoin/bitcoin/pull/20942#discussion_r559495985

ariard commented on src/net_processing.cpp:
```diff
@@ -409,19 +417,33 @@ class PeerManagerImpl final : public PeerManager
+     */
+    bool MarkBlockAsInFlight(CTxMemPool& mempool, NodeId nodeid, const uint256& hash, const CBlockIndex* pindex = nullptr, std::list<QueuedBlock>::iterator** pit = nullptr) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+
+    bool TipMayBeStale(const Consensus::Params &consensusParams) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
a490f0a

"Check if our chain is stalling by measuring if a delta of pow targets has been reached since last time of tip update" ?

ajtowns replied:
Added some text to https://github.com/bitcoin/bitcoin/pull/20758/commits/4f9931df03ac5ea223733abe3909db8c61aa559e

ariard reviewed (commented):
Code Review ACK 6452190, changes are pretty straightforward.

See this stylistic [comment](https://github.com/bitcoin/bitcoin/pull/20942#discussion_r570955961), overall we can order better `PeerManagerImpl` members instead of intertwining methods and vars ? Up to you.

vasild commented:
It would be really nice to get these right before merge:

* Order of member variables and methods, private and public. I think the [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html#Declaration_Order) makes perfect sense.
* Member variables naming.
* Proper doxygen comments.

Otherwise 6452190841f8da1cdaf899d064974136ab37659f looks good.

jnewbery commented:
I think this is probably ready for merge. It's a simple refactor and has three ACKs.

@ajtowns - All three reviewers have asked for sorting members, doxygen comments and renaming data members. I have a small branch that does those things here: https://github.com/jnewbery/bitcoin/tree/pr20942.1. Feel free to take those commits in a follow-up PR.

maflcko reviewed (commented):
Concept ACK 6452190841 I have not reviewed this, but I left a comment 🐡

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Concept ACK 6452190841 I have not reviewed this, but I left a comment 🐡
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiXRwv+Ot8x+JJaOzZ1kM34x9ikAMnoPkV6rbc1xn4csg3qX+vlwOoAsQU9S1VX
UmBGAg9aiY5q5IiNlmBL+17Oay0iQOUYnqkhnPqtoIFHLbTW5QnoC31bi7csjuGj
XASAOnwk8IcXHi19lUdq0LuHld3vdT4bYD+vViO8ANLU8NT7xKgn4/hfXiyL01Qw
rqGGda/3oIzJXZC9/BIBOCfS5xIPfbilQtT3XllyxJBhI8+kblGhbqPQ+GfYZ+to
PoG4mpFzQVrUEL7aYnUapwQ3x0+Rh4+y7UMcz2KC76PjYfDOk1pH8Nnswr8fUBNs
KV7ofcN6cdg6QAdse8GpOQozZdMc706bBLtp1XyZT3MyaRibm/Kw8KYCoQcrqeF2
ZPlzqI55FKybz9YFfAZ0g9xE2G2E/m0k/S2ulbepKboelqvRAQ0j0PMUIV4VQPOW
5QZ0C1ka/jobPQt7XYYxBMhlUpofXzHPe+A0iN6cURPJ9b2SQ+br7BFKAt3aq1HZ
0UwAzlpn
=PF32
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `39fd4a1774698d6853a62e53b467bd6ee8d7c78c218cf8d662c407770c5d971b  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010839fd4a1774698d6853a62e53b467bd6ee8d7c78c218cf8d662c407770c5d971bf01075e680813cb09abaa18bdc00a8c85a3c08fff01007062017e8aa140a665cc171469fe1e508f104602a5484f008b431604c6eca6c0b0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01015358d72024e9e3c60f8386ec9fdc29908f104602a46b1f0080fa91e0cbefbe4e10083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0106d0243b8c600d4263063655586bb841508f104602a5484f008bd5506d77cbc30110083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010f81b230deea2136996b06f9093bfec5708f104602a5483f00899c35bc455da65620083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

amitiuttarwar commented:
fwiw, I had reviewed until a490f0a056456d683dd8ef6f89a5af1a13792118, so post-merge-ACK the first few commits =P

Sjors commented:
Thanks, d44084883adcf00f50d3d5a9e0c88e3a0b276817 allows me to simplify #20295 a bit.
